components/fxa-client/src/jwcrypto.rs

```diff
@@ -0,0 +1,265 @@
+            let mut iv: Vec<u8> = vec![0; 12];
+            rand::fill(&mut iv)?;
+            let nonce = aead::Nonce::try_assume_unique_for_key(encryption_algorithm, &iv)?;
+            let mut encrypted = aead::seal(&sealing_key, nonce, aad, data)?;
+
+            let tag_idx = encrypted.len() - ((128 + 7) >> 3);
+            let tag = encrypted.split_off(tag_idx);
+            let ciphertext = encrypted;
+
+            (None, protected_header, iv, ciphertext, tag)
+        }
+    };
+
+    // Assemble JWE
+    let encrypted_key = encrypted_key
+        .map(|k: Vec<u8>| base64::encode_config(&k, base64::URL_SAFE_NO_PAD))
+        .unwrap_or_default();
+    let header = serde_json::to_string(&protected_header)?;
+    let header = base64::encode_config(header.as_bytes(), base64::URL_SAFE_NO_PAD);
+    let tag = base64::encode_config(auth_tag, base64::URL_SAFE_NO_PAD);
+    let ciphertext = base64::encode_config(ciphertext, base64::URL_SAFE_NO_PAD);
+    let iv = base64::encode_config(iv, base64::URL_SAFE_NO_PAD);
+    Ok(format!(
+        "{}.{}.{}.{}.{}",
+        header, encrypted_key, iv, ciphertext, tag
+    ))
+}
+
+/// Deserialize and decrypt data in the JWE compact form.
+pub fn decrypt_jwe(jwe: &str, encryption_params: DecryptionParameters) -> Result<String> {
+    let segments: Vec<&str> = jwe.split('.').collect();
+    if segments.len() != 5 {
+        return Err(ErrorKind::UnrecoverableServerError("malformed jwe.").into());
+    }
+
+    let protected_header: JweHeader = serde_json::from_slice(&base64::decode_config(
+        &segments[0],
+        base64::URL_SAFE_NO_PAD,
+    )?)?;
+    let encrypted_key = match segments[1].is_empty() {
+        true => None,
+        false => Some(base64::decode_config(
+            &segments[1],
+            base64::URL_SAFE_NO_PAD,
+        )?),
+    };
+    let iv = base64::decode_config(&segments[2], base64::URL_SAFE_NO_PAD)?;
+    let ciphertext = base64::decode_config(&segments[3], base64::URL_SAFE_NO_PAD)?;
+    let auth_tag = base64::decode_config(&segments[4], base64::URL_SAFE_NO_PAD)?;
+
+    match encryption_params {
+        DecryptionParameters::ECDH_ES { local_key_pair } => {
+            if protected_header.alg != Algorithm::ECDH_ES {
+                return Err(ErrorKind::UnrecoverableServerError("alg mismatch.").into());
+            }
+
+            // Part 1: Reconstruct the secret.
+            let peer_jwk = protected_header
+                .epk
+                .as_ref()
+                .ok_or_else(|| ErrorKind::IllegalState("epk not present"))?;
+            let secret = derive_shared_secret(&protected_header, local_key_pair, peer_jwk)?;
+
+            // Part 2: decrypt the payload
+            if encrypted_key.is_some() {
+                return Err(ErrorKind::UnrecoverableServerError(
+                    "The Encrypted Key must be empty.",
+                )
+                .into());
+            }
+            if auth_tag.len() != (128 / 8) {
```

Comment:
As above, you could use the `tag_len` property of the AEAD config itself rather than a constant.